### PR TITLE
Fix IndexError when accessing the "-1" element of dummyarray 

### PR DESCRIPTION
**Pull request by leofang**
This PR fixes the following error:
```python
>>> import numpy as np
>>> from numba import cuda
>>> ary = np.arange(10)
>>> d_ary = cuda.to_device(ary)
>>> d_ary[-1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/cudadrv/devices.py", line 225, in _require_cuda_context
    return fn(*args, **kws)
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/cudadrv/devicearray.py", line 497, in __getitem__
    return self._do_getitem(item)
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/cuda/cudadrv/devicearray.py", line 507, in _do_getitem
    arr = self._dummy.__getitem__(item)
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/dummyarray.py", line 225, in __getitem__
    dims = [dim.__getitem__(it) for dim, it in zip(self.dims, item)]
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/dummyarray.py", line 225, in <listcomp>
    dims = [dim.__getitem__(it) for dim, it in zip(self.dims, item)]
  File "/home/leofang/conda_envs/mpi4py_dev/lib/python3.6/site-packages/numba/dummyarray.py", line 76, in __getitem__
    raise IndexError
IndexError
```

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the PR. Seems like CI is currently failing to due flake8 violations https://dev.azure.com/numba/numba/_build/results?buildId=3744&view=logs&j=687448c2-4012-5c9e-43b3-3520c9635a30&t=dba03d4a-57a7-5e59-a47a-79fe530f297a&l=78

**Comment by leofang** [comment]
@stuartarchibald I silenced it in ba4476a. Is it OK?

**Comment by stuartarchibald** [comment]
> @stuartarchibald I silenced it in [ba4476a](https://github.com/numba/numba/commit/ba4476aed32d5c0fdfb24c1dae61d019d3c60daf). Is it OK?

Could the result just not be assigned to a variable opposed to silencing the error about not assigning the result, or is that not possible here?

**Comment by leofang** [comment]
We could do either of these if you prefer not to do `noqa`:
1. Have a global placeholder `a = None` declared on top, and overwrite it repeatedly, or
2. Take away the assignment and just have a single line like `darr[10]`

I usually don't use the second approach, but it does the job.

**Comment by stuartarchibald** [comment]
> We could do either of these if you prefer not to do `noqa`:
> 
>     1. Have a global placeholder `a = None` declared on top, and overwrite it repeatedly, or
> 
>     2. Take away the assignment and just have a single line like `darr[10]`
> 
> 
> I usually don't use the second approach, but it does the job.

I usually just do 2. Is there some problem you know with this?

**Comment by leofang** [comment]
> I usually just do 2. Is there some problem you know with this?

Pure aesthetic reason 🤣 Let me force push.

**Comment by leofang** [comment]
@stuartarchibald Done. Looks like flake8 is happy now. PTAL, thanks!

**Comment by leofang** [comment]
btw, any chance this fix could go into Numba 0.48 RC?

**Comment by seibert** [comment]
Passed internal CI

File: numba/cuda/tests/cudadrv/test_cuda_array_slicing.py [hunk 1]
```diff
@@ -12,25 +12,49 @@ class CudaArrayIndexing(SerialMixin, unittest.TestCase):
...
+        with self.assertRaises(IndexError):
+            darr[3, 0, 0]
+        with self.assertRaises(IndexError):
+            darr[0, -5, 0]
+        with self.assertRaises(IndexError):
+            darr[0, 4, 0]
+        with self.assertRaises(IndexError):
+            darr[0, 0, -6]
+        with self.assertRaises(IndexError):
+            darr[0, 0, 5]
```
[the remark this answers is not in the document]
Comment (leofang) [reply numba/cuda/tests/cudadrv/test_cuda_array_slicing.py]: @stuartarchibald Done, PTAL.

**Comment by stuartarchibald** [comment]
Thanks for the update, the patch is now failing `flake8`, https://dev.azure.com/numba/numba/_build/results?buildId=3841&view=logs&j=90fddb40-a9bd-5e17-109d-19517de6dba5&t=a765cc85-fe2d-5487-1056-2dc2d92f2a0a&l=78

You may find it useful to `conda install flake8`, then run `flake8 <files you changed>` to check whether it passes locally before pushing. There are also pre-commit hooks available, setup instructions are here:
http://numba.pydata.org/numba-doc/dev/developer/contributing.html#coding-conventions

**Comment by leofang** [comment]
Sorry, I am traveling and do not have access to my laptop, so I edited on the web interface. I will fix it when I can.

**Comment by leofang** [comment]
PTAL

File: numba/cuda/tests/cudadrv/test_cuda_array_slicing.py [hunk 2]
```diff
@@ -12,25 +12,50 @@ class CudaArrayIndexing(SerialMixin, unittest.TestCase):
...
+            darr[x]
 
     def test_index_2d(self):
         arr = np.arange(3 * 4).reshape(3, 4)
         darr = cuda.to_device(arr)
-
-        for i in range(arr.shape[0]):
-            for j in range(arr.shape[1]):
+        x, y = arr.shape
+        for i in range(-x, y):
```
[the remark this answers is not in the document]
Comment (leofang) [reply numba/cuda/tests/cudadrv/test_cuda_array_slicing.py]: Ahh, my bad. Good catch, @stuartarchibald! I just fixed it.

[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/cuda/tests/cudadrv/test_cuda_array_slicing.py]: Great, thanks.

**Comment by seibert** [comment]
btw, I rechecked this on the internal CUDA CI, and it passes.

**Comment by leofang** [comment]
Thanks guys!

**Pull request merged**